kratos/mpi/tests/sources/test_mpi_assign_unique_model_part_collection_tag_utility.cpp

```diff
@@ -0,0 +1,113 @@
+
+        /**
+        * Checks the correct work of the sub modelparts list utility
+        */
+        KRATOS_TEST_CASE_IN_SUITE(AssignMPIUniqueModelPartCollectionTagUtility, KratosMPICoreFastSuite)
+        {
+            // Creating the reference model part and the relative submodelparts non alphabetically ordered
+            auto rank = DataCommunicator::GetDefault().Rank();
+            auto size = DataCommunicator::GetDefault().Size();
+            Model current_model;
+            ModelPart& r_model_part = current_model.CreateModelPart("Main");
+            ModelPart& r_sub_modelpart_1 = r_model_part.CreateSubModelPart("BSubModelPart1");
+            ModelPart& r_sub_modelpart_2 = r_model_part.CreateSubModelPart("ASubModelPart2");
+            ModelPart& r_sub_modelpart_3 = r_model_part.CreateSubModelPart("ZSubModelPart3");
+            ModelPart& r_sub_modelpart_4 = r_model_part.CreateSubModelPart("YSubModelPart4");
+
+            r_model_part.AddNodalSolutionStepVariable(PARTITION_INDEX);
+            CppTestsUtilities::Create2DGeometry(r_model_part, "Element2D3N");
+
+            r_sub_modelpart_1.AddNode(r_model_part.pGetNode(1));
+            r_sub_modelpart_2.AddNode(r_model_part.pGetNode(4));
+            r_sub_modelpart_3.AddNode(r_model_part.pGetNode(5));
+            r_sub_modelpart_4.AddNode(r_model_part.pGetNode(6));
+
+            // Adding nodes to random submodelparts
+            if (DataCommunicator::GetDefault().IsDistributed()) {
+                if (rank == 0) {
+                    r_sub_modelpart_3.AddNode(r_model_part.pGetNode(4));
+                }
+                else {
+                    r_sub_modelpart_3.AddNode(r_model_part.pGetNode(6));
+                    r_sub_modelpart_2.AddNode(r_model_part.pGetNode(5));
+                    r_sub_modelpart_4.AddNode(r_model_part.pGetNode(5));
+
+                }
+            } else {
+                r_sub_modelpart_3.AddNode(r_model_part.pGetNode(4));
+                r_sub_modelpart_3.AddNode(r_model_part.pGetNode(6));
+                r_sub_modelpart_2.AddNode(r_model_part.pGetNode(5));
+                r_sub_modelpart_4.AddNode(r_model_part.pGetNode(5));
+            }
+
+            // Adding elements to random submodelparts
+            r_sub_modelpart_1.AddElement(r_model_part.pGetElement(1));
+            r_sub_modelpart_2.AddElement(r_model_part.pGetElement(4));
+
+
+            AssignUniqueModelPartCollectionTagUtility collections_utility(r_model_part);
+
+            IndexIndexMapType nodes_tags, conds_tags, elems_tags;
+            IndexStringMapType collections;
+            collections_utility.ComputeTags(nodes_tags, conds_tags, elems_tags, collections);
+
+            const std::string filename = "mpi_test";
+
+            AssignUniqueModelPartCollectionTagUtility::WriteTagsToJson(filename + std::to_string(rank), collections);
+            KRATOS_CHECK_EQUAL(collections.size(), 8);
+
+            DataCommunicator::GetDefault().Barrier();
```

Comment:
why barrier?

Reply:
Each rank leaves AssignUniqueModelPartColelctionTagUtility with the coloring, which should be the same. 

In order to check it, I write to file the coloring and then read it to compare the coloring of every rank against all the rest. I put the barrier to ensure that by the time it is reading the files, all ranks have finished writing. Without it, I was afraid that one rank tries to read the file from another rank when it did not finish writing

Reply:
ok sounds good